File: services/speechService.ts
import OpenAI from 'openai';
import { Language } from '../store/useAppStore';

const WHISPER_LANG: Record<Language, string> = {
  zh: 'zh',
  en: 'en',
  ja: 'ja',
};

export async function transcribeAudio(
  audioUri: string,
  language: Language,
  apiKey: string,
): Promise<string> {
  const openai = new OpenAI({ apiKey, dangerouslyAllowBrowser: true });

  const formData = new FormData();
  formData.append('file', {
    uri: audioUri,
    type: 'audio/m4a',
    name: 'recording.m4a',
  } as any);
  formData.append('model', 'whisper-1');
  formData.append('language', WHISPER_LANG[language]);

  const response = await fetch('https://api.openai.com/v1/audio/transcriptions', {
    method: 'POST',
    headers: { Authorization: `Bearer ${apiKey}` },
    body: formData,
  });

  if (!response.ok) throw new Error(`Whisper error: ${response.status}`);
  const data = await response.json();
  return data.text?.trim() ?? '';
}

File: lib/services/speech_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../store/app_store.dart';

Future<String> transcribeAudio({
  required String audioPath,
  required Language language,
  required String apiKey,
}) async {
  final request = http.MultipartRequest(
    'POST',
    Uri.parse('https://api.openai.com/v1/audio/transcriptions'),
  );
  request.headers['Authorization'] = 'Bearer $apiKey';
  request.files.add(await http.MultipartFile.fromPath('file', audioPath));
  request.fields['model'] = 'whisper-1';
  request.fields['language'] = language.whisperCode;

  final streamed = await request.send();
  final body = await streamed.stream.bytesToString();

  if (streamed.statusCode != 200) {
    final err = jsonDecode(body);
    throw Exception(err['error']?['message'] ?? 'Whisper error ${streamed.statusCode}');
  }

  final data = jsonDecode(body);
  return (data['text'] as String).trim();
}
